samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs

```diff
@@ -154,12 +182,63 @@ await foreach (var memory in results)
 
         // Update the token limit.
         memoryText = $"Past memories (format: [memory type] <label>: <details>):\n{memoryText.Trim()}";
-        tokenLimit -= Utils.TokenCount(memoryText);
+        tokenLimit -= Utilities.TokenCount(memoryText);
         context.Variables.Set("tokenLimit", tokenLimit.ToString(new NumberFormatInfo()));
 
         return memoryText;
     }
 
+    /// <summary>
+    /// Extract relevant additional knowledge using a Semantic Kernel planner.
+    /// </summary>
+    [SKFunction("Acquire external information")]
+    [SKFunctionName("AcquireExternalInformation")]
+    [SKFunctionContextParameter(Name = "userIntent", Description = "The intent of the user.")]
+    [SKFunctionContextParameter(Name = "tokenLimit", Description = "Maximum number of tokens")]
+    public async Task<string> AcquireExternalInformationAsync(SKContext context)
+    {
+        if (!this._plannerOptions.Enabled)
+        {
+            return string.Empty;
+        }
+
+        SequentialPlanner planner = await this._plannerFactory(this._kernel);
+
+        // Create a plan and run it.
+        Plan plan = await planner.CreatePlanAsync(context["userIntent"]);
+        while (plan.HasNextStep)
+        {
+            // WORKAROUND for an issue when the planner returns a plan containing a step that is a plan with no steps.
+            // Running these invalid step resets the plan result (plan.State.Input) to the original input/goal statement.
+            var nextStep = plan.Steps[plan.NextStepIndex];
+            if (nextStep.SkillName.Equals("Microsoft.SemanticKernel.Orchestration.Plan", StringComparison.OrdinalIgnoreCase) &&
+                nextStep.Steps.Count == 0)
+            {
+                // Reflect into the plan and bump the next step index - do not try this at home.
+                plan.GetType().GetProperty("NextStepIndex")!.SetValue(plan, plan.NextStepIndex + 1);
+                continue;
+            }
+
+            plan = await planner.Kernel.StepAsync(context.Variables, plan);
```

Comment:
You probably do not want this in the loop as the `context.Variables` will take precedence over the `plan.State`.

```
        // Act

        plan = await this._kernel.StepAsync(context.Variables, plan);

        while (plan.HasNextStep)
        {
            plan = await this._kernel.StepAsync(plan);
        }
```